pkg/util/asyncworker/async_limited_workers.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2022 The Katalyst Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package asyncworker
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+)
+
+func workKeyFunc(obj interface{}) (string, error) {
+	w, ok := obj.(*Work)
+	if !ok {
+		return "", fmt.Errorf("invalid work type")
+	}
+	return w.Name, nil
+}
+
+func workLessFunc(w1, w2 interface{}) bool {
+	return w1.(*Work).DeliveredAt.Before(w2.(*Work).DeliveredAt)
+}
+
+func NewAsyncLimitedWorkers(name string, limit int, emitter metrics.MetricEmitter) *AsyncLimitedWorkers {
```

Comment:
it's better to add a metric to monitor the duration that a work pending in queue (both backoff and wait), if it's complicated we can add a TODO comment and implement it in the future.